Add documentation on how to use one's corefx build

jamesqo opened the pull request:
I forgot how to use the corefx binaries for testing in a console app, after not making PRs for a while. I've added documentation on how to do this in case I ever get stuck on that again, and hopefully prevent others from wasting time.

**To Do**

- [x] Add link to document in `Documentation/README.md`

jamesqo commented on Documentation/debugging/using-your-build.md:
```diff
@@ -0,0 +1,70 @@
+Using Your Build
```
I'm not sure if the `building/` folder would have been a better place for this. I thought of this as more debugging because when I want to use my build in a console app I usually want to debug it, but after writing the document I'm not sure.

jamesqo commented on Documentation/debugging/using-your-build.md:
```diff
@@ -0,0 +1,70 @@
+
+1. Choose a runtime you would like to target your app against.
+
+- A runtime (not to be confused with the [CoreCLR runtime](https://github.com/dotnet/coreclr)) is a specific architecture and OS your .NET app will target.
```
Please comment if you think the wording could be improved here/next bullet point.

jamesqo commented:
/cc @benaadams who originally told me how to do this, @svick, @jnm2

benaadams commented:
/cc @Drawaes who was just asking me about how to do this; so might be dogfooding the instructions 😉

Drawaes commented:
Looking right now 🐶 🍔

petrroll commented on Documentation/debugging/using-your-build.md:
````diff
@@ -0,0 +1,70 @@
+dotnet publish -r <runtime>
+```
+
+4. Replace the stock framework assemblies from `dotnet publish` with your CoreFX build.
````
Replace "from" with something like "produced by" to communicate their origin better?

// Just a nitpick but the "from" made me pause and reread the sentence a few times to understand it.

petrroll commented on Documentation/debugging/using-your-build.md:
````diff
@@ -0,0 +1,70 @@
+```powershell
+# Windows (use PowerShell)
+$source = "<corefx root>/bin/runtime/netcoreapp-<OS>-<configuration>-<architecture>"
+$destination = "<project root>/bin/<configuration>/<framework>/<runtime>/publish"
````
Reference the `dotnetlab` (used in the beginning) to better communicate what the `<project root>` is? E.g. <dotnet lab root> instead?

Just a subjective idea.

jamesqo replied:
I pointed out everything else that needs to be substituted here with angle brackets, so if I don't do it for `<project root>` too, people might forget to edit that part.

mellinoe commented on Documentation/debugging/using-your-build.md:
````diff
@@ -0,0 +1,70 @@
+2. Restore your app's dependencies.
+
+```sh
+dotnet restore -r <runtime>
````
I think the 2.0 SDK does this implicitly when you call "publish", so it might be simpler to just omit this part.

jamesqo replied:
That's a nice improvement, I didn't know the 2.0 SDK did that since I only have 1.0.4 installed. I think it could be frustrating for people who have <2.0 installed to try the instructions, fail and not realize what the problem is, so I will add a note saying `Skip this if you have .NET Core >=2.0 installed.`

mellinoe replied:
It seems a bit fragile to try to drop .NET Core 2.1+ assets (from corefx/master) into a .NET Core 1.1 runtime to me. But it may work in many cases. If you are publishing a .NET Core 2.0 app, then you'll already have the 2.0 toolset which should do the implicit restore.

Not sure how to call that info out. Adding a note like you mentioned seems fine to me.

mellinoe commented on Documentation/debugging/using-your-build.md:
````diff
@@ -0,0 +1,70 @@
+```sh
+mkdir DotnetLab
+cd DotnetLab
+dotnet new
````
It's probably worth mentioning that you should target `netcoreapp2.0` in your "base project" if you intend to copy over the bits from corefx. Technically you want to start with a "base" that is close to what your branch of corefx is producing. For corefx/master that is actually post-netcoreapp2.0, but you should be able to copy master bits on top of 2.0, I think.

jamesqo replied:
@mellinoe I wasn't aware that it mattered. I have a `netcoreapp1.1` app I'm following these instructions with, and things are working fine. Copying the files from corefx's `bin/runtime/` folder overwrites virtually every file in the `publish/` directory, including all of the framework assemblies and the runtime DLLs. So if a file produced by `dotnet publish` won't work with netcoreapp2.0, it'll probably be replaced with one from corefx that will.

mellinoe commented on Documentation/debugging/using-your-build.md:
```diff
@@ -0,0 +1,70 @@
+# Unix
+source=<corefx root>/bin/runtime/netcoreapp-<OS>-<configuration>-<architecture>
+destination=<project root>/bin/<configuration>/<framework>/<runtime>/publish
+cp -rT "$source" "$destination"
```
super-nit: I probably would have written this as `cp $source/* $destination` (similar to the PowerShell version below)

jamesqo replied:
I did this in case `bin/runtime/` includes subdirectories which need to be copied in the future. I have little scripting experience in PS so I assumed `Copy-Item $source/* $dest` would handle recursively copying directories, but it looks like I was wrong. Both versions can probably stay as they are until one of them is broken.

karelz commented:
cc @haralabidis

mellinoe commented:
@jamesqo The latest version looks good as a document by itself.

Not to totally derail the discussion / PR here, but I wonder if we shouldn't try to consolidate this with the [dogfooding.md document](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md)? I was just helping @wfurt with something, and he was referencing the instructions in that other document. It occurs to me now that there's quite a bit of overlap here, especially with the description of how to set up the project. There's sort of two different approaches described, though:

1. Reference the nuget package produced by the corefx build (dogfooding.md)
2. Manually copy the binaries into the published output after the fact (this doc)

Does it make sense to put the non-overlapping information from here over into dogfooding.md?

danmosemsft commented:
There is also coreclr\Documentation\workflow\UsingYourBuild.md which obviously has a different concern but there is overlap. One or the other should describe how to patch both coreclr and corefx at the same time.

PS @jamesqo thanks so much for working on docs!

jamesqo commented:
@mellinoe @danmosemsft Wow, I didn't realize there are already docs in place to deal with this. It makes sense to consolidate the content here with those documents. I'll get to that when I have the time.

jamesqo commented:
@mellinoe I was finally able to make time to consolidate the document with dogfooding.md. I renamed dogfooding to `using-nightly-builds.md` to make the purpose of the document clearer, and also fixed some grammatical issues throughout the document. The meat of the changes are here: https://github.com/jamesqo/corefx/blob/62ccb749511bbe07726f3908177d3cf0d31fefb0/Documentation/project-docs/using-nightly-builds.md#using-your-local-corefx-build

@danmosemsft

> There is also coreclr\Documentation\workflow\UsingYourBuild.md which obviously has a different concern but there is overlap. One or the other should describe how to patch both coreclr and corefx at the same time.

Sorry, I wasn't able to take a look at CoreCLR's docs yet. Perhaps that can be dealt with in a later PR.

mellinoe commented:
@jamesqo You might need to merge in the changes from https://github.com/dotnet/corefx/commit/7d694172b2275ddd0ed1524dbec83f07295672f4.

weshaggard commented:
I'm so used to finding this document by typing dogfooding into the file search can we keep the document called dogfooding.md instead of renaming it?

svick commented on Documentation/project-docs/using-nightly-builds.md:
````diff
@@ -0,0 +1,244 @@
+
+```
+$ dotnet restore
+```
````
Is manually running `dotnet restore` still necessary, since .Net Core SDK 2.0 does it automatically?

I think this also applies to other mentions of `dotnet restore` below.

svick commented on Documentation/project-docs/using-nightly-builds.md:
```diff
@@ -0,0 +1,244 @@
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <RuntimeFrameworkVersion>2.0.0-beta-xyz-00</RuntimeFrameworkVersion> <!-- pick nightly build -->
+    <RuntimeIdentifier>win7-x64</RuntimeIdentifier> <!-- make self-contained -->
```
Isn't the preferred approach on 2.0 to use "portable" RID, i.e. `win-x64`? Though I guess it doesn't matter much for Windows.

mellinoe replied:
Yes, this should say "win-x64".

svick commented on Documentation/project-docs/using-nightly-builds.md:
```diff
@@ -0,0 +1,244 @@
+
+Note #1: There is a [bug](https://github.com/dotnet/sdk/issues/791) with `dotnet run` and self-contained applications. If you `dotnet run` and see an error `The library 'hostpolicy.dll' is required to execute the application...`, you've hit this bug.
+
+Note #2: On non-Windows platforms, self-contained applications aren't runnable by default. You will see an error "Permission denied" when running the application. This is because of a [breaking change in the .NET Core runtime](https://github.com/dotnet/corefx/issues/15516) between 1.0 and 2.0. Either this breaking change needs to be fixed, or [NuGet will have to workaround the change](https://git ...
```
The issues mentioned in both notes above are now closed. Does that mean the two notes no longer apply and should be deleted?

mellinoe replied:
Yeah, I don't think these are relevant anymore.

karelz commented:
What is the status of this PR?
BTW: It might have been easier & faster to just dump the info into our wiki - I will be slowly turning that info into proper docs. It is easier to edit & collaborate, no blocking on reviews ...

karelz commented:
@jamesqo can it be closed now?

Petermarcu commented:
Can we get this one completed? It would be goo to not let it get stale...

@weshaggard dogfooding is such an internal 
Microsoft term, I'm not surprised it wasn't found at first by others. Do you think you could get used to searching for nightly? :)

weshaggard commented:
> Microsoft term, I'm not surprised it wasn't found at first by others. Do you think you could get used to searching for nightly? :)

I'd be fine with that but we will need to update various links to the old dogfooding document across at least corefx and coreclr to have updated pointers.

jamesqo commented:
I will close this until I find the time to get back to it. I would be glad if anyone else is willing to pick this up.

Petermarcu commented:
@karelz or @danmosemsft are one of you able to carry this over the finish line?
